README.md
## What is CacheRules

_CacheRules_ is a well-behaved HTTP caching library aimed at being [RFC 7234](https://tools.ietf.org/html/rfc7234) compliant.

This library does **not actually _cache_ anything**, and it is **not a _proxy_**.
It validates HTTP headers and returns the appropriate response to determine
if a request can be served from the cache.

It is up to the HTTP Cache implementation to store the cached results
and serve responses from the cache if necessary.

[![Build Status](https://travis-ci.org/aw/CacheRules.svg?branch=master)](https://travis-ci.org/aw/CacheRules) [![Coverage Status](https://coveralls.io/repos/aw/CacheRules/badge.svg?branch=master)](https://coveralls.io/r/aw/CacheRules?branch=master) [![Gem Version](https://badge.fury.io/rb/cache_rules.svg)](http://badge.fury.io/rb/cache_rules)

## Getting started

Add this line to your Gemfile: `gem 'cache_rules'`

  or

Install with: `gem install cache_rules`

## Usage

There is only 1 _public API call_ when using this library: `validate()`.

```ruby
require 'cache_rules'

# test without cached response
url     = 'https://status.rubygems.org'
request = {'Version' => 'HTTP/1.1'}
cached  = {}

CacheRules.validate url, request, cached

=> {:body=>nil, :code=>307, :headers=>{"Cache-Lookup"=>"MISS", "Location"=>"https://status.rubygems.org"}}

# test with cached response (status code 200 because no ETag or If-None-Match supplied)
cached  = { "Date" => {"timestamp"=>1420095825}, "X-Cache-Req-Date"  => {"timestamp"=>1420268625}, "X-Cache-Res-Date"  => {"timestamp"=>1420268625} }

CacheRules.validate url, request, cached

=> {:body=>"stale", :code=>200, :headers=>{"Date"=>"Wed, 21 Feb 2018 05:09:27 GMT", "Age"=>"99094242", "Warning"=>"110 - \"Response is Stale\"", "Cache-Lookup"=>"STALE"}}

# test with cached response (status code 304 because If-None-Match supplied)
request = {"Version"=>"HTTP/1.1", "If-None-Match"=>"*"}
cached  = { "Date" => {"timestamp"=>1519190160}, "X-Cache-Req-Date"  => {"timestamp"=>1519190160}, "X-Cache-Res-Date"  => {"timestamp"=>1519190160} }

CacheRules.validate url, request, cached

=> {:body=>nil, :code=>304, :headers=>{"Date"=>"Wed, 21 Feb 2018 05:15:01 GMT", "Age"=>"241", "Warning"=>"110 - \"Response is Stale\"", "Cache-Lookup"=>"STALE"}}
```

The `request` headers must be a Ruby Hash or Array of 2-element Arrays.

The `cached` headers must already have been normalized by this caching library, i.e: they **must** include array keys:

* `Date['timestamp']`
* `X-Cache-Req-Date['timestamp']`
* `X-Cache-Res-Date['timestamp']`

See `test/` directory for more examples.

## Decision tables

There are two decision tables to help figure out how to process each type of
HTTP Caching request.

### Request/Cache Table

![cached](https://cloud.githubusercontent.com/assets/153401/7445031/9d64eb2c-f190-11e4-8938-44a443f9bec3.png)

### Revalidation Table

![revalidation](https://cloud.githubusercontent.com/assets/153401/7445040/09c747ba-f191-11e4-8b9c-a8b26709fb99.png)

## RFC compliance

This HTTP Caching library aims to be RFC 7230-7235 compliant. It is a best
effort attempt to correctly interpret these documents. Some errors may exist,
so please [notify me](https://github.com/aw/CacheRules/issues/new) if something isn't processed correctly according to the RFCs.

### Feature list

  * Normalizing header names and field values (ex: `Last-Modified`)
  * Ensuring date fields are correctly formatted (ex: `Fri, 31 Dec 1999 23:59:59 GMT`)
  * Merging duplicate header fields
  * Interop with HTTP/1.0 clients (ex: `Pragma: no-cache`)
  * Weak entity-tag matching (ex: `If-None-Match: "W/abc123"`)
  * Last modified date matching (ex: `If-Modified-Since: Thu, 01 Jan 2015 07:03:45 GMT`)
  * Various header validation including Cache-Control headers
  * Cache-Control directives with quoted strings (ex: `no-cache="Cookie"`)
  * Removing non-cacheable headers (ex: `Authorization`)
  * Correctly calculating freshness and current age of cached responses
  * Explicit and Heuristic freshness calculation
  * Returning 110 and 111 Warning headers when serving stale responses
  * Revalidating expired responses with the origin server (using HEAD)
  * Returning the correct status code based on validation/revalidation results
  * Lots more little things sprinkled throughout the RFCs...
  * Written in purely functional Ruby (mostly) with 100% unit/integration test coverage

## Custom headers

Custom headers are generated to help with testing and compliance validation.

These are somewhat based on [CloudFlare's](https://support.cloudflare.com/hc/en-us/articles/200168266-What-do-the-various-CloudFlare-cache-responses-HIT-Expired-etc-mean-) cache headers:

  * `Cache-Lookup: HIT`: resource is in cache and still valid. Serving from the cache.
  * `Cache-Lookup: MISS`: resource is not in cache. Redirecting to the origin server.
  * `Cache-Lookup: EXPIRED`: resource is in cache, but expired. Redirecting to the origin server or serving an error message.
  * `Cache-Lookup: STALE`: resource is in cache and expired, but the origin server wasn't contacted successfully to revalidate the request. Serving stale response from the cache.
  * `Cache-Lookup: REVALIDATED`: resource is in cache, was expired, but was revalidated successfully at the origin server. Serving from the cache.

## Tests

To run the tests, type:

  `bundle exec rake test`

## TODO

  * Validation of `s-maxage` [response header](https://tools.ietf.org/html/rfc7234#section-5.2.2.9)
  * Handling `Vary` header and different representations for the same resource
  * Handling `206 (Partial)` and `Range` headers for resuming downloads
  * Handling `Cache-Control: private` headers
  * Caching other _cacheable_ responses such as [404 and 501](https://tools.ietf.org/html/rfc7231#section-6.1)

## What is C.R.E.A.M. ?

C.R.E.A.M. is an influencial lyrical masterpiece from the 90s performed by the [Wu-Tang Clan](https://www.youtube.com/watch?v=PBwAxmrE194)

It's also the premise of this [troll video](https://player.vimeo.com/video/14782834?loop=1;&autoplay=1;)

## Further reading

Some useful articles explaining HTTP Caching:

  * [Caching is hard, draw me a picture](http://www.bizcoder.com/caching-is-hard-draw-me-a-picture)
  * [A Beginner's Guide to HTTP Cache Headers](http://www.mobify.com/blog/beginners-guide-to-http-cache-headers/)

## LICENSE

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at [http://mozilla.org/MPL/2.0/](http://mozilla.org/MPL/2.0/).

Copyright (c) 2014-2018 Alexander Williams, Unscramble


CHANGELOG.md
# Changelog

##  0.6.0 (TBD)

  * Fixes issue #14. Thanks @SleeplessByte
  * Ensure min-fresh cache-control header is used correctly for stale responses

##  0.5.0 (2015-05-04)

  * Add a check which verifies if a precondition exists before revalidating
  * Fixes issue #13
  * Add regression tests
  * Adjust one broken test

##  0.4.1 (2015-05-04)

  * Update README.md

##  0.4.0 (2015-05-03)

  * Modify 'revalidation' table so it doesn't check if the response is expired
  * Ensure validator matching works with '*'
  * Fix tests

##  0.3.0 (2015-05-03)

  * Make cache rules consistent based RFC spec
  * Add required 'cached' headers to tests

##  0.2.0 (2015-05-02)

  * Version bump

##  0.1.20 (2015-05-01)

  * Recently cached responses are cached correctly. Fixes issue #11

##  0.1.19 (2015-05-01)

  * The 'max-stale' header doesn't affect 'max-age' validation

##  0.1.18 (2015-05-01)

  * Ensure the 'Cache-Control: max-age' is validated
  * Add regression tests
  * Update documentation

## 0.1.17 (2015-04-30)

  * Return the proper 'Age' header and ignore the 'public' control directive

## 0.1.16 (2015-04-29)

  * Errors caused by empty HTTP headers. Fixes issue #8

## 0.1.15 (2015-04-29)

  * No max-stale = allow anything. Fixes issue #7

## 0.1.14 (2015-04-28)

  * Ensure Cache-Control is verified

## 0.1.13 (2015-04-28)

  * Don't cache the Host header

## 0.1.12 (2015-04-28)

  * Allow newer rubies

## 0.1.11 (2015-04-20)

  * Remove Gemfile.lock
  * Adjust Gemfile dependencies
  * Run travis-ci tests in container

## 0.1.10 (2015-04-20)

  * Remove gemspec post_install message

## 0.1.9 (2015-04-08)

  * Add regression tests for issue #5
  * Update Gemfile.lock / dependencies
  * Fix dates in this CHANGELOG

## 0.1.8 (2015-02-09)

  * Add tests to ensure URLs with query parameters are maintained

## 0.1.7 (2015-02-02)

  * Refactor and simplify `revalidate_response()` method

## 0.1.6 (2015-02-02)

  * Don't rescue ArgumentError on httpdate parse errors

## 0.1.5 (2015-02-02)

  * Closes #3. Returns all cached headers according to RFC 7234 sec4.3.4

## 0.1.4 (2015-02-01)

  * HTTP `Age` header is now returned as a String, but processed as an Integer


test/helper.rb
require 'coveralls'
require 'fakeweb'
require 'simplecov'

Coveralls.wear!

SimpleCov.start do
  add_filter '/test/'
  add_filter '.bundle'
  minimum_coverage 100
end

require './lib/cache_rules'
require 'minitest/autorun'
require 'minitest/unit'
require 'minitest/reporters'

Minitest::Reporters.use! [Minitest::Reporters::SpecReporter.new]


test/test_tables.rb
class TestTables < MiniTest::Test

  def setup
    @request_map  = {"0000000"=>0, "0000001"=>0, "0000010"=>1, "0000011"=>1, "0000100"=>0, "0000101"=>0, "0000110"=>1, "0000111"=>1, "0001000"=>0, "0001001"=>0, "0001010"=>1, "0001011"=>1, "0001100"=>0, "0001101"=>0, "0001110"=>1, "0001111"=>1, "0010000"=>0, "0010001"=>0, "0010010"=>1, "0010011"=>1, "0010100"=>0, "0010101"=>0, "0010110"=>1, "0010111"=>1, "0011000"=>0, "0011001"=>0, "0011010"=>1, "0011011"=>1, "0011100"=>0, "0011101"=>0, "0011110"=>1, "0011111"=>1, "0100000"=>0, "0100001"=>0, "0100010"=>1, "0100011"=>1, "0100100"=>0, "0100101"=>0, "0100110"=>1, "0100111"=>1, "0101000"=>0, "0101001"=>0, "0101010"=>1, "0101011"=>1, "0101100"=>0, "0101101"=>0, "0101110"=>1, "0101111"=>1, "0110000"=>0, "0110001"=>0, "0110010"=>1, "0110011"=>1, "0110100"=>0, "0110101"=>0, "0110110"=>1, "0110111"=>1, "0111000"=>0, "0111001"=>0, "0111010"=>1, "0111011"=>1, "0111100"=>0, "0111101"=>0, "0111110"=>1, "0111111"=>1, "1000000"=>2, "1000001"=>2, "1000010"=>2, "1000011"=>2, "1000100"=>6, "1000101"=>4, "1000110"=>6, "1000111"=>4, "1001000"=>3, "1001001"=>3, "1001010"=>3, "1001011"=>3, "1001100"=>6, "1001101"=>5, "1001110"=>6, "1001111"=>5, "1010000"=>8, "1010001"=>8, "1010010"=>8, "1010011"=>8, "1010100"=>8, "1010101"=>8, "1010110"=>8, "1010111"=>8, "1011000"=>8, "1011001"=>8, "1011010"=>8, "1011011"=>8, "1011100"=>8, "1011101"=>8, "1011110"=>8, "1011111"=>8, "1100000"=>2, "1100001"=>2, "1100010"=>2, "1100011"=>2, "1100100"=>7, "1100101"=>7, "1100110"=>7, "1100111"=>7, "1101000"=>3, "1101001"=>3, "1101010"=>3, "1101011"=>3, "1101100"=>7, "1101101"=>7, "1101110"=>7, "1101111"=>7, "1110000"=>8, "1110001"=>8, "1110010"=>8, "1110011"=>8, "1110100"=>8, "1110101"=>8, "1110110"=>8, "1110111"=>8, "1111000"=>8, "1111001"=>8, "1111010"=>8, "1111011"=>8, "1111100"=>8, "1111101"=>8, "1111110"=>8, "1111111"=>8}
    @response_map = {"000"=>0, "001"=>1, "010"=>0, "011"=>1, "100"=>2, "101"=>2, "110"=>3, "111"=>4}
  end

  def test_x_value
    assert_kind_of NilClass, CacheRules::X

    assert_nil CacheRules::X
  end

  def test_request_table
    assert_kind_of Hash, CacheRules::REQUEST_TABLE

    assert_includes CacheRules::REQUEST_TABLE, :conditions
    assert_includes CacheRules::REQUEST_TABLE, :actions

    assert_equal CacheRules::REQUEST_TABLE[:conditions].length, 7
    assert_equal CacheRules::REQUEST_TABLE[:actions].length, 6

    conditions = CacheRules::REQUEST_TABLE[:conditions].keys
    actions    = CacheRules::REQUEST_TABLE[:actions].keys
    assert_equal conditions, %w(cached must_revalidate no_cache precond_match expired only_if_cached allow_stale)
    assert_equal actions,    %w(revalidate add_age add_x_cache add_warning add_status return_body)
  end

  def test_response_table
    assert_kind_of Hash, CacheRules::RESPONSE_TABLE

    assert_includes CacheRules::RESPONSE_TABLE, :conditions
    assert_includes CacheRules::RESPONSE_TABLE, :actions

    assert_equal CacheRules::RESPONSE_TABLE[:conditions].length, 3
    assert_equal CacheRules::RESPONSE_TABLE[:actions].length, 6

    conditions = CacheRules::RESPONSE_TABLE[:conditions].keys
    actions    = CacheRules::RESPONSE_TABLE[:actions].keys
    assert_equal conditions, %w(is_error allow_stale validator_match)
    assert_equal actions,    %w(revalidate add_age add_x_cache add_warning add_status return_body)
  end

  def test_request_map_is_correct
    assert_kind_of Hash, CacheRules::REQUEST_MAP

    assert_equal CacheRules::REQUEST_MAP.length, 128
    assert_equal CacheRules::REQUEST_MAP, @request_map

    assert_equal CacheRules::REQUEST_MAP["0000000"], 0
    assert_equal CacheRules::REQUEST_MAP["0000010"], 1
    assert_equal CacheRules::REQUEST_MAP["1000000"], 2
    assert_equal CacheRules::REQUEST_MAP["1001000"], 3
    assert_equal CacheRules::REQUEST_MAP["1000101"], 4
    assert_equal CacheRules::REQUEST_MAP["1001101"], 5
    assert_equal CacheRules::REQUEST_MAP["1000100"], 6
    assert_equal CacheRules::REQUEST_MAP["1100100"], 7
    assert_equal CacheRules::REQUEST_MAP["1010000"], 8
  end

  def test_response_map_is_correct
    assert_kind_of Hash, CacheRules::RESPONSE_MAP

    assert_equal CacheRules::RESPONSE_MAP.length, 8
    assert_equal CacheRules::RESPONSE_MAP, @response_map

    assert_equal CacheRules::RESPONSE_MAP["000"], 0
    assert_equal CacheRules::RESPONSE_MAP["001"], 1
    assert_equal CacheRules::RESPONSE_MAP["100"], 2
    assert_equal CacheRules::RESPONSE_MAP["110"], 3
    assert_equal CacheRules::RESPONSE_MAP["111"], 4
  end

end


lib/actions.rb
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Copyright (c) 2014-2016 Alexander Williams, Unscramble <license@unscramble.jp>

module CacheRules
  extend self

  # HTTP Header Actions

  def action_revalidate(result)
    result[:value]
  end

  # Generate an age equal to the current cached entry's age
  # source: https://tools.ietf.org/html/rfc7234#section-4
  def action_add_age(result)
    current_age = helper_current_age Time.now.gmtime.to_i, result[:cached]

    {'Age' => current_age.to_s}
  end

  def action_add_x_cache(result)
    {'Cache-Lookup' => result[:value]}
  end

  def action_add_warning(result)
    {'Warning' => result[:value]}
  end

  def action_add_status(result)
    result[:value]
  end

  def action_return_body(result)
    result[:value]
  end

end


lib/formatting.rb
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Copyright (c) 2014-2016 Alexander Williams, Unscramble <license@unscramble.jp>

module CacheRules
  extend self

  # HTTP Header Formatting

  # Create a normalized Hash of HTTP headers
  def normalize
    ->(headers) {
      Hash[normalize_fields.(combine.(clean.(Array(headers).map &format_key)))]
    }
  end

  # Format the key to look like this: Last-Modified
  def format_key
    Proc.new {|key, value|
      k = key.downcase == 'etag' ? 'ETag' : key.split('-').map(&:capitalize).join('-')
      [ k, value ]
    }
  end

  # Intentionally drop these headers to avoid caching them
  # If-Modified-Since should be dropped if the date isn't valid
  # source: https://tools.ietf.org/html/rfc7232#section-3.3
  def clean
    ->(headers) {
      Array(headers).reject {|key, value|
        HEADERS_NO_CACHE.include?(key) || helper_is_if_modified_error?(key, value) || value.nil? || value.empty?
      }
    }
  end

  # Combine headers with a comma if the field-names are duplicate
  def combine
    ->(headers) {
      Array(headers).group_by {|h, _| h }.map {|k, v|
        v = HEADERS_CSV.include?(k) ? v.map {|_, x| x }.join(', ') : v[0][1] # OPTIMIZE
        [ k, v ]
      }
    }
  end

  # Normalizes the value (field-value) of each header
  def normalize_fields
    ->(headers) {
      Array(headers).map &format_field
    }
  end

  # Returns a Hash of Strings
  def unnormalize_fields
    ->(headers) {
      Array(headers).reduce({}) {|hash, (key, value)|
        hash.merge Hash[[format_field.call(key, value, true)]]
      }
    }
  end

  # Returns a Hash, Array, Integer or String based on the supplied arguments
  def format_field
    Proc.new {|key, header, stringify|
      f = format_value header, stringify

      value = case key
        when *HEADERS_HTTPDATE then f.call('httpdate')      # => Hash
        when *OPTIONS_CACHE    then f.call('cache_control') # => Array
        when *OPTIONS_CSV      then f.call('csv')           # => Array
        when *OPTIONS_RETRY    then f.call('retry_after')   # => Hash or Integer
        else header                                         # => String
      end
      [ key, value ]
    }
  end

  # Returns the value of the field
  def format_value(header, stringify = nil)
    Proc.new {|field|
      stringify ? send("#{ field }_string", header) : send("#{ field }", header)
    }
  end

  def httpdate(header)
    timestamp = httpdate_helper header

    {
      'httpdate'  => Time.at(timestamp).gmtime.httpdate,
      'timestamp' => timestamp
    }
  end

  def httpdate_string(header)
    timestamp = httpdate_helper header['httpdate']

    Time.at(timestamp).gmtime.httpdate
  end

  # Correctly parse the 3 Date/Time formats and convert to GMT
  # source: https://tools.ietf.org/html/rfc7234#section-4.2
  def httpdate_helper(header)
    # source: https://tools.ietf.org/html/rfc7231#section-7.1.1.1
    DateTime.parse(header).to_time.to_i
  rescue => e
    # If the supplied date is invalid, use a time in the past (5 minutes ago)
    # source: https://tools.ietf.org/html/rfc7234#section-5.3
    Time.now.gmtime.to_i - 300
  end

  # OPTIMIZE: this regex is copied from JavaScript, could be greatly simplified
  # Returns a Hash with the directive as key, token (or nil), quoted-string (or nil)
  def cache_control(header = '')
    result = header.scan /(?:^|(?:\s*\,\s*))([^\x00-\x20\(\)<>@\,;\:\\"\/\[\]\?\=\{\}\x7F]+)(?:\=(?:([^\x00-\x20\(\)<>@\,;\:\\"\/\[\]\?\=\{\}\x7F]+)|(?:\"((?:[^"\\]|\\.)*)\")))?/
    result.reduce({}) {|hash, x|
      hash.merge({
        x[0].downcase => {
          'token'         => x[1],
          'quoted_string' => x[2]
        }
      })
    }
  end

  # Parses the Cache-Control header and returns a comma-separated String
  def cache_control_string(header)
    Array(header).map {|x|
      token     = x[1]['token']
      quote     = x[1]['quoted_string']
      directive = x[0]

      if    token && quote.nil? then "#{ directive }=#{ token }"
      elsif token.nil? && quote then "#{ directive }=\"#{ quote }\""
      else  directive
      end
    }.join ', '
  end

  def csv(header = '')
    header.split(',').map(&:strip)
  end

  def csv_string(header)
    Array(header).join ', '
  end

  # "The value of this field can be either an HTTP-date or a number of seconds..."
  # source: https://tools.ietf.org/html/rfc7234#section-7.1.3
  def retry_after(header)
    Integer(header).abs
  rescue => e
    httpdate header
  end

  def retry_after_string(header)
    header.to_s
  end

end


lib/validations.rb
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Copyright (c) 2014-2016 Alexander Williams, Unscramble <license@unscramble.jp>

module CacheRules
  extend self

  # HTTP Header Validators
  #
  # Parameters must always be normalized
  #
  # Return must always be 0 or 1
  #
  # The If-Match and If-Unmodified-Since conditional header fields are not applicable to a cache.
  # source: https://tools.ietf.org/html/rfc7234#section-4.3.2

  def to_bit(&predicate)
    predicate.call ? 1 : 0
  end

  def validate_cached?(headers)
    to_bit { headers[:cached].length > 0 }
  end

  # Precedence: If-None-Match (ETag), then If-Modified-Since (Last-Modified)
  # source: https://tools.ietf.org/html/rfc7232#section-6
  def validate_precond_match?(headers)
    request, cached = headers.values_at :request, :cached
    return 0 if cached.length == 0

    # Return when the If-None-Match header exists, ignore If-Modified-Since
    # source: https://tools.ietf.org/html/rfc7232#section-3.3
    etag_match = helper_etag(request, cached)
    return to_bit { etag_match } unless etag_match.nil?

    to_bit { helper_last_modified(request, cached) }
  end

  # Compare headers to see if the cached request is expired (Freshness)
  # source: https://tools.ietf.org/html/rfc7234#section-4.2
  def validate_expired?(headers)
    return 0 if headers[:cached].length == 0

    freshness_lifetime, current_age = helper_freshness_lifetime.call headers[:cached]

    response_is_fresh = freshness_lifetime.to_i > current_age

    return 1 if headers[:cached]['Cache-Control'] &&
                  headers[:cached]['Cache-Control']['max-age'] &&
                    current_age > headers[:cached]['Cache-Control']['max-age']['token'].to_i


    to_bit { (response_is_fresh != true) }
  end

  def validate_only_if_cached?(headers)
    to_bit { headers[:request]['Cache-Control'] && headers[:request]['Cache-Control']['only-if-cached'] }
  end

  # Serving Stale Responses
  # source: https://tools.ietf.org/html/rfc7234#section-4.2.4
  def validate_allow_stale?(headers)
    request, cached = headers.values_at :request, :cached
    return 0 if cached.length == 0 || helper_validate_allow_stale(request, cached)

    freshness_lifetime, current_age = helper_freshness_lifetime.call cached

    max_stale = helper_max_stale.call request['Cache-Control'], freshness_lifetime, current_age

    to_bit { max_stale || max_stale.nil? }
  end

  # Response Cache-Control Directives
  # source: https://tools.ietf.org/html/rfc7234#section-5.2.2
  def validate_must_revalidate?(headers)
    return 1 if headers[:cached].length == 0

    # source: https://tools.ietf.org/html/rfc7234#section-5.2.2.1
    # source: https://tools.ietf.org/html/rfc7234#section-5.2.2.7
    to_bit { (( cached = headers[:cached]['Cache-Control'] )) && ( cached['must-revalidate'] || cached['proxy-revalidate'] ) }
  end

  # Verify if we're explicitly told not to serve a response without revalidation
  def validate_no_cache?(headers)
    request_headers, cached_headers = headers.values_at :request, :cached
    return 1 if cached_headers.length == 0

    # Must revalidate if this request header exists
    # source: https://tools.ietf.org/html/rfc7234#section-5.2.1.4
    return 1 if (request_headers['Cache-Control'] && request_headers['Cache-Control']['no-cache'])

    freshness_lifetime, current_age = helper_freshness_lifetime.call cached_headers

    # Revalidate if the cached response is still fresh and will be fresh for at least min_fresh seconds
    # see GitHub issue #14
    # source: https://tools.ietf.org/html/rfc7234#section-5.2.1.3
    return 1 if helper_min_fresh.call request_headers['Cache-Control'], freshness_lifetime, current_age

    # If max-age is 0 or if the current age is above the max-age
    # source: https://tools.ietf.org/html/rfc7234#section-5.2.1.1
    return 1 if request_headers['Cache-Control'] &&
                  request_headers['Cache-Control']['max-age'] &&
                    (request_headers['Cache-Control']['max-age']['token'].to_s == "0" || current_age > request_headers['Cache-Control']['max-age']['token'].to_i)

    # source: https://tools.ietf.org/html/rfc7234#section-5.2.2.2
    # source: https://tools.ietf.org/html/rfc7234#section-3.2
    if cached_headers['Cache-Control']
      return 1 if (( cached = cached_headers['Cache-Control'] )) &&
        helper_no_cache.call(cached_headers)                               ||
          (cached['no-cache'] && cached['no-cache']['quoted_string'].nil?) ||
          (cached['s-maxage'] && cached['s-maxage']['token'].to_s == "0")  ||
          (cached['max-age'] && cached['max-age']['token'].to_s   == "0")
    end

    # source: https://tools.ietf.org/html/rfc7234#section-5.4
    # Legacy support for HTTP/1.0 Pragma header
    return 1 if request_headers['Pragma'] && request_headers['Pragma']['no-cache']

    return 0
  end

  def validate_is_error?(headers)
    to_bit { headers[:response]['Status'].to_i.between?(500,599) }
  end

  def validate_validator_match?(headers)
    request, response = headers.values_at :request, :response
    to_bit { response['ETag'] && request['If-None-Match'] && (request['If-None-Match'].include?(response['ETag']) || request['If-None-Match'].include?("*")) }
  end

end


lib/cache_rules.rb
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Copyright (c) 2014-2016 Alexander Williams, Unscramble <license@unscramble.jp>
#
# Original Source:  https://github.com/aw/CacheRules
#
#
# This library validates requests and responses for cached HTTP data.
#
# Rules based on RFCs 7230-7235 - https://tools.ietf.org/wg/httpbis/
#
# Usage:
# => CacheRules.validate(url, request_headers, cached_headers)
#
# If you grew up on the crime side:
# => CacheRules::Everything::Around::Me.validate(url, request_headers, cached_headers)

require 'net/http'
require 'date'
require 'time'
require 'uri'

require 'actions.rb'
require 'formatting.rb'
require 'helpers.rb'
require 'validations.rb'

module CacheRules
  extend self

  HEADERS_NO_CACHE = %w(
    Set-Cookie Cookie
    Accept-Ranges Range If-Range Content-Range
    Referer From Host
    Authorization Proxy-Authorization
  )

  HEADERS_HTTPDATE = %w(
    Last-Modified If-Modified-Since If-Unmodified-Since
    Expires Date
    X-Cache-Req-Date X-Cache-Res-Date
  )

  HEADERS_CSV = %w(
    Connection Trailer Transfer-Encoding Upgrade Via
    Accept Accept-Charset Accept-Encoding Accept-Language Allow
    Content-Encoding Content-Language Vary
    Cache-Control Warning Pragma If-Match If-None-Match
  )

  HEADERS_NUMBER = %w(
    Age Content-Length Max-Forwards
  )

  OPTIONS_CACHE = HEADERS_CSV.select {|header| header == 'Cache-Control' }
  OPTIONS_CSV   = HEADERS_CSV.reject {|header| header == 'Cache-Control' }
  OPTIONS_RETRY = %w(Retry-After)

  X = nil

  # Decision table for request/cached headers
  REQUEST_TABLE = {
    :conditions => {
      'cached'          => [0, 0, 1, 1, 1, 1, 1, 1, 1],
      'must_revalidate' => [X, X, X, X, 0, 0, 0, 1, X],
      'no_cache'        => [X, X, 0, 0, 0, 0, 0, 0, 1],
      'precond_match'   => [X, X, 0, 1, 0, 1, X, X, X],
      'expired'         => [X, X, 0, 0, 1, 1, 1, 1, X],
      'only_if_cached'  => [0, 1, X, X, X, X, X, X, X],
      'allow_stale'     => [X, X, X, X, 1, 1, 0, X, X]
    },
    :actions => {
      'revalidate'      => [X, X, X, X, X, X, X, 1, 1],
      'add_age'         => [X, X, 1, 1, 1, 1],
      'add_x_cache'     => %w(MISS MISS HIT HIT STALE STALE EXPIRED),
      'add_warning'     => [X, X, X, X, '110 - "Response is Stale"', '110 - "Response is Stale"'],
      'add_status'      => [307, 504, 200, 304, 200, 304, 504],
      'return_body'     => [X, 'Gateway Timeout', 'cached', X, 'stale', X, 'Gateway Timeout']
    }
  }

  # Decision table for revalidated responses
  RESPONSE_TABLE = {
    :conditions => {
      'is_error'        => [0, 0, 1, 1, 1],
      'allow_stale'     => [X, X, 0, 1, 1],
      'validator_match' => [0, 1, X, 0, 1]
    },
    :actions => {
      'revalidate'      => [],
      'add_age'         => [1, 1, X, 1, 1],
      'add_x_cache'     => %w(REVALIDATED REVALIDATED EXPIRED STALE STALE),
      'add_warning'     => [X, X, X, '111 - "Revalidation Failed"', '111 - "Revalidation Failed"'],
      'add_status'      => [200, 304, 504, 200, 304],
      'return_body'     => ['cached', X, 'Gateway Timeout', 'stale']
    }
  }

  # Build the map tables in advance for faster lookups i.e: O(1)
  REQUEST_MAP   = helper_table_map(REQUEST_TABLE[:conditions])
  RESPONSE_MAP  = helper_table_map(RESPONSE_TABLE[:conditions])

  # Public: Validate a URL and the request/cached/response headers
  # TODO: validate the required parameters to ensure they are set correctly
  def validate(url, request_headers, cached_headers = {})
    # 1. normalize the request headers
    normalized_headers = normalize.call request_headers
    actions            = REQUEST_TABLE[:actions]

    # 2. get the column matching the request headers
    column = REQUEST_MAP[helper_run_validate.call(REQUEST_TABLE[:conditions], normalized_headers, cached_headers).join]
    response    = Proc.new { helper_response url, actions, column, cached_headers }
    revalidate  = Proc.new { revalidate_response url, normalized_headers, cached_headers }

    # 3. return the response or revalidate
    actions['revalidate'][column] == 1 ? revalidate.call : response.call
  end

  # Revalidates a response by fetching headers from the origin server
  def revalidate_response(*args)
    url, request, cached = *args
    has_preconditions    = helper_has_preconditions.(request, cached)

    # 1. get the column
    column = if has_preconditions
      res_headers = helper_response_headers.(helper_make_request_timer.(args))
      RESPONSE_MAP[helper_run_validate.call(RESPONSE_TABLE[:conditions], request, cached, res_headers).join]
    else
      res_headers = {}
      2 # return column 2 (504 EXPIRED)
    end

    # 2. return the response
    helper_response url, RESPONSE_TABLE[:actions], column, cached, res_headers
  rescue => error
    {:code => 504, :body => 'Gateway Timeout', :headers => [], :error => error.message, :debug => error}
  end

  # Returns a net/http response Object
  def make_request
    ->(url, request_headers, cached_headers) {
      uri               = URI.parse url
      http              = Net::HTTP.new uri.host, uri.port
      http.open_timeout = 2
      http.read_timeout = 60
      http.use_ssl      = uri.scheme == 'https'
      http.verify_mode  = OpenSSL::SSL::VERIFY_PEER

      request           = Net::HTTP::Head.new uri.request_uri

      # Two possible validators: entity tags and timestamp
      # source: https://tools.ietf.org/html/rfc7234#section-4.3.1
      entity_tags = Proc.new { helper_combine_etags request_headers, cached_headers }.call
      timestamp   = Proc.new { helper_timestamp     request_headers, cached_headers }

      # Set the precondition header before making the request
      request['If-None-Match'] = entity_tags if entity_tags
      ts = timestamp.call unless entity_tags
      request['If-Modified-Since'] = ts if ts && !entity_tags

      # Make the HTTP(s) request
      helper_make_request http, request
    }
  end

end

module CacheRules
  module Everything
    module Around
      module Me
        extend CacheRules

        # C.R.E.A.M.
        def self.get_the_money
          "Dolla Dolla Bill Y'all"
        end

      end
    end
  end
end
